### PR TITLE
Test for cluster add remove scenario

### PR DESCRIPTION
**Pull request by Jaanki**
This defines a structure for a test scenario
where a new cluster is added to an existing setup
and connectivity between pods in the new and
existing cluster is tested.

Signed-Off-By: Janki Chhatbar <jchhatba@redhat.com>

---
This change is [Reviewable](https://reviewable.io/reviews/submariner-io/submariner/202)

### CODE REVIEW

**Comment by Jaanki** [comment]
Dependant on #198

**Comment by Jaanki** [comment]
@tpantelis @mangelajo can you guys take a look plz :)

File: test/e2e/add_cluster.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    "github.com/submariner-io/submariner/test/e2e/framework"
+
+    . "github.com/onsi/ginkgo"
+)
+
+var _ = PDescribe("[example] Test to check adding a new cluster to existing setup", func() {
+    f := framework.NewDefaultFramework("add-cluster")
+
+    // Start listener pod1 on cluster 1
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply test/e2e/add_cluster.go]: Done

File: test/e2e/framework/nodes.go [hunk 2]
```diff
@@ -9,26 +9,27 @@ import (
...
+func (f *Framework) FindNodesByLabel(cluster ClusterIndex) (gateway []*v1.Node, nonGateway []*v1.Node) {
 	nodes := AwaitUntil("list nodes", func() (interface{}, error) {
 		// Ignore the control plane node labeled as master as it doesn't allow scheduling of pods
 		return f.ClusterClients[cluster].CoreV1().Nodes().List(metav1.ListOptions{
 			LabelSelector: "!node-role.kubernetes.io/master",
 		})
 	}, NoopCheckResult).(*v1.NodeList)
 
-	var gatewayName, nonGatewayName string
+	var gatewayNodes, nonGatewayNodes []*v1.Node
```
Comment (Jaanki) [inline test/e2e/framework/nodes.go]: @tpantelis This is what I was suggesting yesterday.

File: test/e2e/dataplane/tcp_pod_to_service.go [hunk 3]
```diff
@@ -14,45 +14,54 @@ var _ = Describe("[dataplane] Basic Pod to Service tests across clusters without
...
-	listenerPod, connectorPod := runAndVerifyNetworkPod2ServicePair(f, rightScheduling, leftScheduling)
+	listenerPod, connectorPod := runAndVerifyNetworkPod2ServicePair(f, rightScheduling, leftScheduling, framework.ClusterB, framework.ClusterA)
 
 	framework.Logf("Connector pod has IP: %s", connectorPod.Pod.Status.PodIP)
 	By("Verifying the output of listener pod which must contain the source IP")
 	Expect(listenerPod.TerminationMessage).To(ContainSubstring(connectorPod.Pod.Status.PodIP))
 }
 
-func runAndVerifyNetworkPod2ServicePair(f *framework.Framework, leftScheduling framework.NetworkPodScheduling, rightScheduling framework.NetworkPodScheduling) (*framework.NetworkPod, *framework.NetworkPod) {
+func TestPod2ServiceTCPNoConnection(f *framework.Framework, leftScheduling framework.NetworkPodScheduling, rightScheduling framework.NetworkPodScheduling, listernerPodCluster framework.ClusterIndex, connectorPodCluster framework.ClusterIndex) {
```
Comment (tpantelis) [inline test/e2e/dataplane/tcp_pod_to_service.go]: listernerPodCluster -> listenerPodCluster

File: test/e2e/dataplane/tcp_pod_to_service.go [hunk 4]
```diff
@@ -14,45 +14,54 @@ var _ = Describe("[dataplane] Basic Pod to Service tests across clusters without
...
+	//TODO Determine what to "Expect"
+	Expect("Error to have occured")
+}
+
+func runAndVerifyNetworkPod2ServicePair(f *framework.Framework, leftScheduling framework.NetworkPodScheduling, rightScheduling framework.NetworkPodScheduling, listernerPodCluster framework.ClusterIndex, connectorPodCluster framework.ClusterIndex) (*framework.NetworkPod, *framework.NetworkPod) {
 	By("Creating a listener pod in cluster B, which will wait for a handshake over TCP")
 	listenerPod := f.NewNetworkPod(&framework.NetworkPodConfig{
 		Type:       framework.ListenerPod,
-		Cluster:    framework.ClusterB,
+		Cluster:    listernerPodCluster, //framework.ClusterB,
```
Comment (tpantelis) [inline test/e2e/dataplane/tcp_pod_to_service.go]: Please remove the comment

File: test/e2e/dataplane/tcp_pod_to_service.go [hunk 5]
```diff
@@ -63,7 +72,7 @@ func runAndVerifyNetworkPod2ServicePair(f *framework.Framework, leftScheduling f
 	By("Creating a connector pod in cluster A, which will attempt the specific UUID handshake over TCP")
 	connectorPod := f.NewNetworkPod(&framework.NetworkPodConfig{
 		Type:       framework.ConnectorPod,
-		Cluster:    framework.ClusterA,
+		Cluster:    connectorPodCluster, //framework.ClusterA,
```
Comment (tpantelis) [inline test/e2e/dataplane/tcp_pod_to_service.go]: here too

File: test/e2e/dataplane/tcp_pod_to_service.go [hunk 6]
```diff
@@ -14,45 +14,54 @@ var _ = Describe("[dataplane] Basic Pod to Service tests across clusters without
...
+func TestPod2ServiceTCPNoConnection(f *framework.Framework, leftScheduling framework.NetworkPodScheduling, rightScheduling framework.NetworkPodScheduling, listernerPodCluster framework.ClusterIndex, connectorPodCluster framework.ClusterIndex) {
+
+	runAndVerifyNetworkPod2ServicePair(f, leftScheduling, rightScheduling, listernerPodCluster, connectorPodCluster)
+
+	By("Verifying that the connection attempt from connector pod errors out")
+	//TODO Determine what to "Expect"
+	Expect("Error to have occured")
+}
+
+func runAndVerifyNetworkPod2ServicePair(f *framework.Framework, leftScheduling framework.NetworkPodScheduling, rightScheduling framework.NetworkPodScheduling, listernerPodCluster framework.ClusterIndex, connectorPodCluster framework.ClusterIndex) (*framework.NetworkPod, *framework.NetworkPod) {
```
Comment (tpantelis) [inline test/e2e/dataplane/tcp_pod_to_service.go]: listernerPodCluster -> listenerPodCluster

File: test/e2e/dataplane/tcp_pod_to_service.go [hunk 7]
```diff
@@ -14,45 +14,54 @@ var _ = Describe("[dataplane] Basic Pod to Service tests across clusters without
...
 	})
 
 	It("Should preserve the source IP (NonGW to NonGW node)", func() {
-		testPod2ServiceTCPIPPreservation(f, framework.NonGatewayNode, framework.NonGatewayNode)
+		TestPod2ServiceTCPIPPreservation(f, framework.NonGatewayNode, framework.NonGatewayNode)
 	})
 })
 
-func testPod2ServiceTCP(f *framework.Framework, leftScheduling framework.NetworkPodScheduling, rightScheduling framework.NetworkPodScheduling) {
+func TestPod2ServiceTCP(f *framework.Framework, leftScheduling framework.NetworkPodScheduling, rightScheduling framework.NetworkPodScheduling, listernerPodCluster framework.ClusterIndex, connectorPodCluster framework.ClusterIndex) {
```
Comment (tpantelis) [inline test/e2e/dataplane/tcp_pod_to_service.go]: Can you make these signature changes in a separate PR as I'll need it as well?

File: test/e2e/framework/framework.go [hunk 8]
```diff
@@ -275,3 +294,50 @@ func createNamespace(client kubeclientset.Interface, name string, labels map[str
 	Expect(err).NotTo(HaveOccurred(), "Error creating namespace %v", namespaceObj)
 	return namespace
 }
+
+func DoPatchOperation(path string, value string, patchFunc PatchFunc) {
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply test/e2e/framework/framework.go]: Not part of this PR...

File: test/e2e/framework/framework.go [hunk 9]
```diff
@@ -275,3 +294,50 @@ func createNamespace(client kubeclientset.Interface, name string, labels map[str
...
+	AwaitUntil("perform patch operation", func() (interface{}, error) {
+		return nil, patchFunc(types.JSONPatchType, payloadBytes)
+	}, NoopCheckResult)
+}
+
+func NoopCheckResult(interface{}) (bool, error) {
+	return true, nil
+}
+
+func AwaitUntil(opMsg string, doOperation DoOperationFunc, checkResult CheckResultFunc) interface{} {
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply test/e2e/framework/framework.go]: Not part of this PR...

File: test/e2e/framework/nodes.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+			gateway = &nodes.Items[i]
+			gatewayName = gateway.Name
+		} else {
+			nonGateway = &nodes.Items[i]
+			nonGatewayName = nonGateway.Name
+		}
+	}
+
+	Logf("Found gateway node: %s, non-gateway node: %s", gatewayName, nonGatewayName)
+	return
```
[the remark this answers is not in the document]
Comment (tpantelis) [reply test/e2e/framework/nodes.go]: Not part of this PR...

File: test/e2e/dataplane/tcp_pod_to_service.go (hunk 3 above)
Comment (Jaanki) [reply test/e2e/dataplane/tcp_pod_to_service.go]: Done

File: test/e2e/dataplane/tcp_pod_to_service.go (hunk 6 above)
Comment (Jaanki) [reply test/e2e/dataplane/tcp_pod_to_service.go]: Done

File: test/e2e/dataplane/tcp_pod_to_service.go (hunk 5 above)
Comment (Jaanki) [reply test/e2e/dataplane/tcp_pod_to_service.go]: Done

File: test/e2e/dataplane/tcp_pod_to_service.go (hunk 4 above)
Comment (Jaanki) [reply test/e2e/dataplane/tcp_pod_to_service.go]: Done

File: test/e2e/dataplane/tcp_pod_to_service.go (hunk 7 above)
Comment (Jaanki) [reply test/e2e/dataplane/tcp_pod_to_service.go]: https://github.com/submariner-io/submariner/pull/205

File: test/e2e/add_cluster.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package e2e
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply test/e2e/add_cluster.go]: Done. Used add_remove_cluster.go

**Comment by Jaanki** [comment]
Depends on #205

**Comment by Jaanki** [comment]
Closing in favor of #235

**Pull request closed without merging**